site/content/blog/2019-10-21-localizing-matterpoll/index.md

```diff
@@ -0,0 +1,182 @@
+
+```go
+// getServerLocalizer returns a localizer that localizes in the server default client locale
+func (p *MatterpollPlugin) getServerLocalizer() *i18n.Localizer {
+	return i18n.NewLocalizer(p.bundle, *p.ServerConfig.LocalizationSettings.DefaultClientLocale)
+}
+```
+
+With this preparation we where finally ready to do the actual task: Defining and translating actual strings. Putting the pieces together is quite simple and comes down to:
+
+```go
+	l := p.getUserLocalizer(userID)
+	response, err := l.LocalizeMessage(&i18n.Message{
+		ID:    "response.vote.counted",
+		Other: "Your vote has been counted.",
+	}
+	
+	// Send response back
+```
+
+Most translation strings can be translated without any context information. But for example the "Total Votes" text you saw above include the number of votes. We did incorporate this information directly into the `Message`:
+
+```go
+	l := p.getUserLocalizer(userID)
+	response, err := l.Localize(&i18n.LocalizeConfig{
+		DefaultMessage: &i18n.Message{
+			ID:    "poll.message.totalVotes",
+			Other: "**Total votes**: {{.TotalVotes}}",
+		},
+		TemplateData:   map[string]interface{}{"TotalVotes": numberOfVotes},
+	})
+		
+	// Send response back
+```
+
+### Extracting the translation strings
+
+As already mentioned [`goi18n`](https://github.com/nicksnyder/go-i18n#command-goi18n-) allow the extraction of translation string from the code into the translation file. This can be done by running
+`goi18n extract -format json -outdir assets/i18n/ server/`, which creates `assets/i18n/active.en.json` containing all translation strings.
+
+#### Adding a new language
+
+If we e.g. want to add support for german, which has the language code `de`, we have to:
+
+1. Create it's own file: `touch assets/i18n/translate.de.json`
+- Merge all translation strings into the new file: `goi18n merge -format json -outdir assets/i18n/ assets/i18n/active.en.json assets/i18n/translate.de.json`
+- Translate all strings in `translate.de.json`
+- Rename it from `translate.de.json` to `active.de.json`.
+
+#### Translating new strings
+
+1. Extract the new goi18n translations strings: `goi18n extract -format json -outdir assets/i18n/ server/`
+- Update your translation file: `goi18n merge -format json -outdir assets/i18n/ assets/i18n/active.*.json`
+- Translate all strings for a language e.g. in `active.de.json`
+- Merge the translations in the active files: `goi18n merge -format json -outdir assets/i18n/ assets/i18n/active.*.json assets/i18n/translate.de.json`
+
+### Future Ideas
+
+This is a rough outline how plugins can use the existing framework to fully support localization. If other plugin developers adapt this approach, it might get officially supported by Mattermost. This means that the plugin framework would support translations via go-i18n.
```

Comment:
```suggestion
This is a rough outline how plugins can use the existing framework to fully support localization. If other plugin developers adapt this approach, it might get officially supported by Mattermost. This means that the plugin framework would support translations via `go-i18n`.
```

Reply:
See https://github.com/mattermost/mattermost-developer-documentation/pull/421/files#r344622433